raiden/utils/gevent_utils.py

```diff
@@ -14,15 +14,90 @@ class RaidenGreenlet(gevent.Greenlet):
     """
     def link_safe(self, callable):
         self.is_safe_linked = True
-        super(RaidenGreenlet, self).link(self._wrap_linked_callable(callable))
+        super().link(self._wrap_linked_callable(callable))
+
+    def link_exception_safe(self, callable):
+        self.is_safe_linked = True
+        super().link_exception(self._wrap_linked_callable(callable))
+
+    def rawlink_safe(self, callable):
+        self.is_safe_linked = True
+        super().rawlink(self._wrap_linked_callable(callable))
+
+    @staticmethod
+    def _wrap_linked_callable(callable):
+        def wrapped_callable(greenlet):
+            try:
+                callable(greenlet)
+            except gevent.get_hub().SYSTEM_ERROR:
+                raise
+            except Exception as exception:
+                raise UnhandledExceptionInGreenlet(exception) from exception
+        return wrapped_callable
+
+
+class RaidenGreenlet(gevent.Greenlet):
+    def link_safe(self, callable):
+        self.is_safe_linked = True
+        super().link(self._wrap_linked_callable(callable))
+
+    def link_exception_safe(self, callable):
+        self.is_safe_linked = True
+        super().link_exception(self._wrap_linked_callable(callable))
+
+    def rawlink_safe(self, callable):
+        self.is_safe_linked = True
+        super().rawlink(self._wrap_linked_callable(callable))
+
+    @staticmethod
+    def _wrap_linked_callable(callable):
+        def wrapped_callable(greenlet):
+            try:
+                callable(greenlet)
+            except gevent.get_hub().SYSTEM_ERROR:
+                raise
+            except Exception as exception:
+                raise UnhandledExceptionInGreenlet(exception) from exception
+        return wrapped_callable
+
+
+class RaidenAsyncResult(gevent.event.AsyncResult):
+    def link_safe(self, callable):
+        self.is_safe_linked = True
+        super().link(self._wrap_linked_callable(callable))
+
+    def link_exception_safe(self, callable):
+        self.is_safe_linked = True
+        super().link_exception(self._wrap_linked_callable(callable))
+
+    def rawlink_safe(self, callable):
+        self.is_safe_linked = True
+        super().rawlink(self._wrap_linked_callable(callable))
+
+    @staticmethod
+    def _wrap_linked_callable(callable):
```

Comment:
Can't you:
```
def _wrap_linked_callable(callable):
    @wraps(callable)
    def wrapped_callable(greenlet):
        try:
            callable(greenlet)
        except gevent.get_hub().SYSTEM_ERROR:
            raise
        except Exception as exception:
            raise UnhandledExceptionInGreenlet(exception) from exception
    return wrapped_callable

def patch_link_method(cls, method_name):
    unsafe = getattr(cls, method_name)
    def safe_link(self, callable):
        return unsafe(self, _wraps_linked_callable(callable))
    setattr(cls, method_name, safe_link)

patch_link_method(gevent.event.AsyncResult, 'link')
patch_link_method(gevent.event.AsyncResult, 'link_exception')
...
```
? That should work for Greenlet and friends too

Reply:
I think these gevent classes are builtin and can't be patched. This is why we introduced a RaidenGreenlet instead of just patching Greenlet...

Reply:
These gevent classes (at least `Greenlet`) are extension types that cannot be patched. That's why we introduced `RaidenGreenlet` in the first place...

Reply:
@jomuel Right, but the other classes are standard python classes, and even for `Greenlet`, I suppose a inherited class with same name as a import-time drop-in replacement would require less refactoring, something in the lines:
```
from gevent import Greenlet as GGreenlet

class Greenlet(GGreenlet):
    is_safe_linked = True
    def link(self, callable):
        return super().link(_wrap_link_callable(callable)
    ...
```
then just `from raiden.utils.gevent_utils import Greenlet` instead of importing from `gevent`, and no other changes to the calls should be required, no?